### ISSUE TITLE
Migrate to `ruff` for linting and formatting

### ISSUE THREAD

**Issue by sphuber** [aiidateam/aiida-core#6178]
The [`ruff`](https://docs.astral.sh/ruff/) tool started out as a Python linter with the goal to improve the speed of existing linters such as `pylint`. It has been seeing a huge uptake in the Python community because it is orders of magnitude faster. I have been holding off on migrating to it since it was in rapid development and didn't have quite the level of coverage of rules as `pylint` does. However, that has improved significantly and now it also provides a formatter which is just as fast. Switching to `ruff` would allow us to drop `isort`, `yapf`, `pylint` and `pydocstyle` and have everything in one superfast tool. I have done the migration on `aiida-shell` and it has been buttery-smooth. I would propose to do the same for `aiida-core` (and probably all other Python packages in `aiidateam`).

**Comment by unkcpz** [comment]
I'd also propose to look at type checking in rust family https://github.com/mtshiba/pylyzer, I NEVER get mypy test all passed locally, even through we didn't cover all codes.

**Comment by sphuber** [comment]
> I'd also propose to look at type checking in rust family https://github.com/mtshiba/pylyzer, I NEVER get mypy test all passed locally, even through we didn't cover all codes.

I find it a bit weird that they don't even mention `mypy` in their README when they compare themselves to other tools. That is by far the industry standard. They have a really good team working on it, including Guido himself, so I have quite a lot of trust in their code. I personally have no problems with running mypy locally. Are you sure you have the correct version installed in your environment? The pre-commit uses the version of `mypy` installed in the local env, and doesn't create a separate pre-commit env for it. So you have to make sure the version you have installed is the version that is mentioned in the `pre-commit` extra of our install dependencies in `pyproject.toml`.

**Comment by unkcpz** [comment]
I think https://pyre-check.org/ is more like an alternative/competitor to `mypy` and it is maintained by meta (facebook).

**Comment by unkcpz** [comment]
I get a change to run `pre-commit run --all` from a brand new environment. I got following errors.

```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

aiida/engine/processes/functions.py:47: error: Module has no attribute "UnionType"; maybe "FunctionType"?  [attr-defined]
aiida/engine/processes/functions.py:50: error: Unused "type: ignore[assignment, misc]" comment  [unused-ignore]
aiida/engine/processes/functions.py:53: error: Module "typing" has no attribute "ParamSpec"; maybe "_ParamSpec"?  [attr-defined]
aiida/engine/processes/functions.py:53: note: Use `from typing_extensions import ParamSpec` instead
aiida/engine/processes/functions.py:53: note: See https://mypy.readthedocs.io/en/stable/runtime_troubles.html#using-new-additions-to-the-typing-module
aiida/engine/processes/functions.py:56: error: Unused "type: ignore" comment  [unused-ignore]
aiida/engine/processes/functions.py:59: error: Module has no attribute "get_annotations"; maybe "__annotations__"?  [attr-defined]
aiida/engine/processes/functions.py:102: error: Free type variable expected in Generic[...]  [misc]
aiida/engine/processes/functions.py:105: error: Name "P.args" is not defined  [name-defined]
aiida/engine/processes/functions.py:105: error: Name "P.kwargs" is not defined  [name-defined]
aiida/engine/processes/functions.py:108: error: Name "P.args" is not defined  [name-defined]
aiida/engine/processes/functions.py:108: error: Name "P.kwargs" is not defined  [name-defined]
aiida/engine/processes/functions.py:111: error: Name "P.args" is not defined  [name-defined]
aiida/engine/processes/functions.py:111: error: Name "P.kwargs" is not defined  [name-defined]
aiida/engine/processes/functions.py:114: error: Name "P.args" is not defined  [name-defined]
aiida/engine/processes/functions.py:114: error: Name "P.kwargs" is not defined  [name-defined]
aiida/engine/processes/functions.py:128: error: The first argument to Callable must be a list of types, parameter specification, or "..."  [valid-type]
aiida/engine/processes/functions.py:128: note: See https://mypy.readthedocs.io/en/stable/kinds_of_types.html#callable-types-and-lambdas
aiida/engine/processes/functions.py:128: error: "ProcessFunctionType" expects 2 type arguments, but 3 given  [type-arg]
aiida/engine/processes/functions.py:128: error: Variable "aiida.engine.processes.functions.P" is not valid as a type  [valid-type]
aiida/engine/processes/functions.py:128: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
aiida/engine/processes/functions.py:152: error: Unused "type: ignore" comment  [unused-ignore]
aiida/engine/processes/functions.py:155: error: The first argument to Callable must be a list of types, parameter specification, or "..."  [valid-type]
aiida/engine/processes/functions.py:155: note: See https://mypy.readthedocs.io/en/stable/kinds_of_types.html#callable-types-and-lambdas
aiida/engine/processes/functions.py:155: error: "ProcessFunctionType" expects 2 type arguments, but 3 given  [type-arg]
aiida/engine/processes/functions.py:155: error: Variable "aiida.engine.processes.functions.P" is not valid as a type  [valid-type]
aiida/engine/processes/functions.py:155: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
aiida/engine/processes/functions.py:179: error: Unused "type: ignore" comment  [unused-ignore]
aiida/storage/sqlite_zip/migrations/versions/main_0000a_replace_nulls.py:46: error: Unused "type: ignore" comment  [unused-ignore]
aiida/storage/sqlite_zip/migrations/versions/main_0000a_replace_nulls.py:47: error: Unused "type: ignore" comment  [unused-ignore]
aiida/storage/sqlite_zip/migrations/versions/main_0000a_replace_nulls.py:64: error: Unused "type: ignore" comment  [unused-ignore]
aiida/storage/sqlite_zip/migrations/versions/main_0000a_replace_nulls.py:65: error: Unused "type: ignore" comment  [unused-ignore]
Found 26 errors in 2 files (checked 754 source files)

pylint...................................................................Passed
Update all requirements files............................................Passed
Validate environment.yml.................................................Passed
Automatically generating verdi docs......................................Passed
```

My `mypy` and `pre-commit` versions:

```
(base) aiida@48b56a403499:~/aiida-core$ pip freeze | grep mypy
mypy==1.7.1
mypy-extensions==1.0.0
(base) aiida@48b56a403499:~/aiida-core$ pip freeze | grep pre-commit
pre-commit==2.21.0
```

The `aiida-core` is installed from the latest main (https://github.com/aiidateam/aiida-core/commit/5015f5fe12d93024ed0d7594d860f1f2cd977548) and installed by `pip install -U -e ".[tui,docs,atomic_tools,pre-commit,tests,rest]"`

**Comment by unkcpz** [comment]
Am I missing some packages to run it without problem? In my other dev environments I also cannot run `pre-commit run` locally.

**Comment by sphuber** [comment]
Are you perhaps running with Python 3.9 or older?

**Comment by unkcpz** [comment]
Yes, with python3.9.

**Comment by sphuber** [comment]
That would explain at the very least some of the errors. Can you try running it with Python 3.11? I am pretty sure that many, if not all warnings will disappear

**Comment by unkcpz** [comment]
thanks @sphuber.
Test with `python==3.10.13` the mypy error down to four, I guess now it worth a quick fix for them?

```
aiida/storage/sqlite_zip/migrations/versions/main_0000a_replace_nulls.py:46: error: Unused "type: ignore" comment  [unused-ignore]
aiida/storage/sqlite_zip/migrations/versions/main_0000a_replace_nulls.py:47: error: Unused "type: ignore" comment  [unused-ignore]
aiida/storage/sqlite_zip/migrations/versions/main_0000a_replace_nulls.py:64: error: Unused "type: ignore" comment  [unused-ignore]
aiida/storage/sqlite_zip/migrations/versions/main_0000a_replace_nulls.py:65: error: Unused "type: ignore" comment  [unused-ignore]
```

Back to the issue itself, the `pylint` takes most of the time for pre-commit, I support it 200 percents.

**Comment by sphuber** [comment]
I don't think we need a fix, because the CI is passing just fine. So I think there is simply still a difference in y9our environment and that used on GHA. It uses Python 3.10 though, so maybe it is a difference in `sqlalchemy` or `alembic`? Those libraries are implicated in those lines.

**Comment by sphuber** [comment]
And yes, `pylint` is super slow. I have already prepared the branch to migrate to ruff. But that was a few weeks ago. Plan is to merge it after the release is out, but that is taking longer than expected.

**Comment by unkcpz** [comment]
Still see the same errors with python3.11, I am curious what those errors are, so I open a draft https://github.com/aiidateam/aiida-core/pull/6230, if the pre-commit still fine, I think we can remove those.

**Comment by unkcpz** [comment]
Found the root in https://github.com/sqlalchemy/alembic/issues/1058, it is the `alembic` version which is pinned in `requirements/requiremnts-py-3.10.txt` as `1.11.1`. Now the `mypy` issue is fixed in `alembic==1.13` which is installed with aiida-core.

**Comment by sphuber** [comment]
Fixed in #6233
